### ISSUE TITLE
Overwrite Executable Account Data/Lamports during Tx

### ISSUE THREAD

**Issue by Reisen** [solana-labs/solana#10390]
#### Problem

It is possible to manipulate smart contract execution for the duration of a single transaction. This would allow me to replace BPF contract ELFs for the duration of a transaction and allow me to manipulate accounts owned by that contract.

##### Cause

1. When loading executable accounts [[1]], any executable owner will also be loaded in a loop. By creating accounts owned by executable accounts, `MessageProcessor` will receive a list of `executable_accounts` > 1.
2. The message processor will prepend all of these accounts [[2]] to the keyed account list passed to instruction processors. Due to the assumption in process_instruction that the first address will be the executing program, a list of `[1..]` will include these extra executable accounts and expose them to the processor.
3. When verifying post-instruction data only `pre_accounts` and `accounts` are checked. Neither of these include the list of executable transactions prepended by step 2 and so are never processed by `verify` [[3]].
4. This allows changes to the executable accounts passed to instructions to persist from one instruction to the next. Instruction 1 could replace a BPF program in an executable used by instruction two, which can then do as it pleases with any owned accounts.

[1]: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts.rs#L209
[2]: https://github.com/solana-labs/solana/blob/master/runtime/src/message_processor.rs#L277
[3]: https://github.com/solana-labs/solana/blob/master/runtime/src/message_processor.rs#L486

#### Proposed Solution

@jackcmay Already has a local fix I believe.

**Comment by jackcmay** [comment]
@Reisen  Were you able to demonstrate the attack you are describing?

**Comment by mvines** [comment]
The BPFLoader allows instruction 1 to rewrite the list of accounts that is then passed by the MessageProcessor to instruction 2?   That seems really unlikely to me, so it sounds like this attack would be limited to native programs

**Comment by Reisen** [comment]
@jackcmay Only through playing with the test I originally showed you, as I can't do it on TdS without some way of creating an executable account. I'll see if I can fix up the script for the above and paste it here, perhaps I missed something.

@mvines So It's not the BPFLoader specifically. It's the MessageProcessor which appends executable accounts to _any_ program. Unless I misunderstood, It's actually native programs which are safe.

**Comment by jackcmay** [comment]
@Reisen If I remember correctly you test relied on direct access to the bank in order to add a new loader (chained programs).  I'd love to see this re-created on `devnet`

**Comment by Reisen** [comment]
Right! Yeah so the direct access in the test was just to do the chaining of the programs because the BPF loader is the only way to create programs, but didn't have this facility yet. That was the main reason for sitting on this the past couple of weeks and leaving the bug fix with you, to see if there were any OTHER ways of parenting an executable account currently.

I don't think it can be done on devnet directly right now until that facility is unlocked, and I stopped looking. I opened this issue to describe the main attack that would occur and also to unblock your local bug fix.
